src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -643,55 +625,161 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                // Also, even if we're not dealing with an enum, we still create a new array type that drops any
+                // annotations that may have initially been associated with the element type; this is similarly to
+                // ensure deterministic behavior.
+                arrayType = arrayType.WithElementType(TypeWithAnnotations.Create(elementType.EnumUnderlyingTypeOrSelf()));
+
+                var cachingField = _builder.module.GetArrayCachingFieldForData(data, _module.Translate(arrayType), wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                var arrayNotNullLabel = new object();
+
+                // T[]? array = PrivateImplementationDetails.cachingField;
+                // if (array is not null) goto arrayNotNull;
+                _builder.EmitOpCode(ILOpCode.Ldsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitBranch(ILOpCode.Brtrue, arrayNotNullLabel);
+
+                // array = new T[elementCount];
+                // RuntimeHelpers.InitializeArray(token, array);
+                // PrivateImplementationDetails.cachingField = array;
+                _builder.EmitOpCode(ILOpCode.Pop);
+                _builder.EmitIntConstant(elementCount);
+                _builder.EmitOpCode(ILOpCode.Newarr);
+                EmitSymbolToken(arrayType.ElementType, wrappedExpression.Syntax);
+                _builder.EmitArrayBlockInitializer(data, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitOpCode(ILOpCode.Stsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+
+                // arrayNotNullLabel:
+                // new ReadOnlySpan<T>(array)
+                _builder.MarkLabel(arrayNotNullLabel);
+                _builder.EmitOpCode(ILOpCode.Newobj, 0);
+                EmitSymbolToken(rosArrayCtor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                return true;
+            }
+
+            // Emit: new ReadOnlySpan<ElementType>(PrivateImplementationDetails.cachingField ??= new ElementType[] { ... constants ... })
+            bool tryEmitAsCachedArrayOfConstants(BoundArrayCreation arrayCreation, ArrayTypeSymbol arrayType, TypeSymbol elementType, NamedTypeSymbol spanType, bool used, BoundExpression? inPlaceTarget, out bool avoidInPlace)
+            {
+                avoidInPlace = false;
+                var initializer = arrayCreation.InitializerOpt;
+                if (initializer == null)
+                {
+                    return false;
+                }
+
+                var initializers = initializer.Initializers;
+                if (initializers.Any(static init => init.ConstantValueOpt == null))
```

Comment:
>init.ConstantValueOpt == null [](http://example.com/codeflow?start=52&length=29)

I am not sure if caching an array of strings is such a good idea. This will hold on to strings forever, at the same time this could be the only time they are used.

Reply:
Tagging @stephentoub for thoughts

Reply:
> This will hold on to strings forever, at the same time this could be the only time they are used.

As string literals, won't they already be held on to forever as part of interning?

Reply:
> As string literals, won't they already be held on to forever as part of interning?

To be honest, I do not know the answer. And whether the answer is the same of all flavors of frameworks out there.

> this doesn't seem different to me from, say, lambda/delegate caching

Strings could be quite big though. And there could be a lot of them in a single initialization. Also, we do not cache delegates when they are created by using `new`. So, there is some control over that form of caching.

Reply:
> To be honest, I do not know the answer. And whether the answer is the same for all flavors of frameworks out there.

It should be true for both .NET Framework and .NET Core, every time we encounter a string literal we add it to a global hashmap (string interning Stephen mentioned above) where it's essentially rooted forever. Except the cases with unloadable ALCs but I guess it should not be a problem here as well. E.g.

```cs
void Test(bool cond)
{
    if (cond)
        Console.WriteLine("true!!");
    else
        Console.WriteLine("false!!");
}
```
When JIT compiles this method (on its first execution) it will create string objects for both literals even if one of them (e.g. `false!!!`) will never be used - we might make it more efficient in future, but it's a current behavior of .NET Framework and .NET Core.

Reply:
@stephentoub @dotnet/roslyn-compiler From discussion with Aleksey, we still have some concern about permanent allocation even for scenarios that might only have one usage. This concern applies to the new types covered by this PR, but also the previous types in cases where an allocation is involved. Aleksey proposed we use a WeakReference and I think it's a good idea. 
Please chime in if you have concerns. I'll move ahead with that proposal.

Reply:
> Please chime in if you have concerns. I'll move ahead with that proposal.

I have concerns, especially about "but also the previous types in cases where an allocation is involved".   You're saying for a case like https://github.com/dotnet/runtime/blob/7f8da5754e382fcc5d0992a3f72b17d4ee1e9b05/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCaseEquivalences.Data.cs#L170 (which is compiled into the regex source generator which targets netstandard2.0) this might end up allocating a new 20Kb array each time it's accessed if a GC occurs in between?

Reply:
In addition, I still don't fully understand the concern. Let's say you have this method:
```C#
ReadOnlySpan<nint> M()
{
    nint[] arr = new nint[1000];
    arr[0] = 0;
    arr[1] = 1;
    ...
    arr[998] = 998;
    arr[999] = 999;
    return arr;
}
```
I believe the concern being expressed is this will end up caching an ~8K array (on 64-bit), even if it might not be used again, yes? But the IL for this method, which is also going to be loaded even if never used again, is ~12K, and the assembly code generated for it in tier 0 (which is what you'll get the first time it's executed) is ~32K (then if it gets promoted to tier 1, another ~10K will be generated), so we've got at least ~44K of memory dedicated to this method whether or not we incur the additional ~8K for the array. Has that been factored into your concerns?

Reply:
> Also, we do not cache delegates when they are created by using new. So, there is some control over that form of caching.

That same control is available here as well if you really want it. If you write:
```C#
int[] arr = new int[] { ... };
```
it won't be cached.

Reply:
> That same control is available here as well if you really want it.

I do not think the control that you are talking about is intuitive. I would even say that the behavior change caused by extraction of a local, or an inline of a local, might come as a big surprise. When different language constructs have different behavior, that is a different story, and that is what we have with delegate caching.

Reply:
> Regarding RegexCaseEquivalences.Data.cs. That code could be written to explicitly perform unconditional strong caching, if that is desired. I do not think compiler should be the one making a decision like that. Simply because the caching is appropriate for one specific component, it doesn't mean that it is appropriate for everyone else, I think.

We would need to revisit every single use of this in dotnet/runtime where the source might be compiled downlevel, as this could regress all of them.  Are you planning to do that?

Reply:
>  I would even say that the behavior change caused by extraction of a local, or an inline of a local, might come as a big surprise.

There are plenty of situations where that's the case, including in the brand new collection expressions feature.

Reply:
> We would need to revisit every single use of this in dotnet/runtime where the source might be compiled downlevel, as this could regress all of them. Are you planning to do that?

If we decide that the behavior change, which, If remember correctly, was introduced without much discussion at the time (and likely specifically for the benefit of a single component in development at the time) was a mistake, and should be changed, then we will have to decide what to do with the component.

Reply:
> There are plenty of situations where that's the case, including in the brand new collection expressions feature.

Doesn't mean it is a good thing. The nature of differences is not the same, and the impact is quite different. Each situation is somewhat unique.

Reply:
I agree there is a non-zero chance this causes unexpected behavior for a customer. There are definitely customers out there that generate large `string[]` for initialization purposes that are effectively single use. I cannot specifically remember a case where it would combined with `ReadOnlySpan<string>` such that it would trigger this optimization but it's certainly possible. 

There are other optimizations we've taken in the past that had the potential to negatively impact customer scenarios. Even simple optimizations like increased method group to delegate caching broke partner teams. It is always going to be a trade off. 

The criteria I usually consider is:

1. Is this on the whole going to be an improvement? In this case I believe the answer is yes it's overall going to produce significant wins compared to the potential downside. 
2. Is there a reasonable and documented way the user can _undo_ the optimization if it's found to be negative? Consider as an example for method group caching the _undo_ operation was just make the delegate allocation explicit: `= new Action(Method)` vs. `= Method`. What is the _undo_ operation here? I believe assigning to an intermediate local would subvert the optimization. Is that the way we want to document? Whatever the answer is I would like for it to be explicitly listed in the PR / issue for customers to see. 
3. Are we violating anything in the language specification? For method group to delegate allocation we had to go back and confirm with LDM that they were okay with this change. 

Assuming we have resolutions for (2) and (3) I would overall be in favor of moving forward. I tihnk we should consider an entry in the breaking change list though.

Reply:
Was asked if I could make it clearer which optimization path I preferred: weak reference or not. 

I would lean towards starting the non-weak reference approach. My rationale is 

1. Customers who find the behavior undesirable can use the _undo_ mechanism 
2. If this does produce enough negative customer feedback we could flip to the `WeakReference` approach in an update / servicing fix.